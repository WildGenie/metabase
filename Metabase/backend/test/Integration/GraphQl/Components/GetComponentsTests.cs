using System.Threading.Tasks;
using Snapshooter.Xunit;
using Xunit;
using System.Linq;
using FluentAssertions;
using System.Collections.Generic;

namespace Metabase.Tests.Integration.GraphQl.Components
{
    [Collection(nameof(Data.Component))]
    public sealed class GetComponentsTests
      : ComponentIntegrationTests
    {
        [Fact]
        public async Task NoComponent_ReturnsEmptyList()
        {
            // Act
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }

        [Fact]
        public async Task SingleComponent_IsReturned()
        {
            // Arrange
            await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            var componentId = await CreateComponentReturningId(MinimalComponentInput).ConfigureAwait(false);
            LogoutUser();
            // Act
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                matchOptions => matchOptions.Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.components.edges[*].node.id").Should().Be(componentId)
                 )
                );
        }

        [Fact]
        public async Task MultipleComponents_AreReturned()
        {
            // Arrange
            await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            var componentIds = new List<string>();
            foreach (var input in ComponentInputs)
            {
                componentIds.Add(
                    await CreateComponentReturningId(input).ConfigureAwait(false)
                    );
            }
            LogoutUser();
            // Act
            var response = await GetComponents().ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                matchOptions =>
                    componentIds.Select((componentId, index) => (componentId, index)).Aggregate(
                        matchOptions,
                        (accumulatedMatchOptions, componentIdAndIndex) =>
                            accumulatedMatchOptions.Assert(fieldOptions =>
                                fieldOptions.Field<string>($"data.components.edges[{componentIdAndIndex.index}].node.id").Should().Be(componentIdAndIndex.componentId)
                                )
                )
            );
        }
    }
}