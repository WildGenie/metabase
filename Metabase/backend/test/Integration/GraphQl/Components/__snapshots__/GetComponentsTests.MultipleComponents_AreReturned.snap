{"data":{"components":{"edges":[{"node":{"id":"Q29tcG9uZW50Cmc3ZWNkM2IwZjdjMDc0MDBhOWRiM2Q1ODdlMDc4YmM4Nw==","name":"Component A","abbreviation":"C!A","description":"Best component ever!","availability":null,"categories":[]}},{"node":{"id":"Q29tcG9uZW50CmdjYjMwMjI4ZTA4Zjk0ZTNiYjY4NTA1ODA1NWYyY2I0Mw==","name":"Component B","abbreviation":"C!B","description":"Another component!","availability":{"from":"2020-01-01T08:00:00.000Z","to":"2020-11-05T20:00:00.000Z"},"categories":["MATERIAL","UNIT"]}},{"node":{"id":"Q29tcG9uZW50CmdjNTE5ODNiYTVmNGM0OGI0ODQ1NDk0ZjI5YmM0ZjFjMg==","name":"Component C","abbreviation":"C!C","description":"Yet another component!","availability":{"from":null,"to":"2021-11-05T20:00:00.000Z"},"categories":["UNIT"]}},{"node":{"id":"Q29tcG9uZW50CmdhOWYxY2NiMGU4YjQ0ZmE0OGRlZjEwYmRhMWE1YjU1Mw==","name":"Component D","abbreviation":"C!D","description":"Whatever component!","availability":{"from":"2019-04-03T00:00:00.000Z","to":null},"categories":["LAYER","MATERIAL","UNIT"]}}]}}}
